File: Code/transactions.cpp
#include <iostream>
#include <string>
#include <fstream>
#include <algorithm>
#include <cstring>
#include <unordered_map>

#include "menus.h"
#include "user_file_process.h"
#include "games_file_process.h"
#include "constants.h"
#include "utils.h"
#include "transactions.h"


using namespace std;



transactions::transactions()
{

}

// logic for sell transaction
string transactions::process_sell(vector<string>& all_games,vector<string>& games_to_add, string current_user_name,vector<string>& transaction_log)
{

    string new_game_name;
    cout<<"Sell"<<endl;
    cout<<"Enter game name:";



    cin.ignore();
    getline(cin, new_game_name);
    new_game_name=utils().pad_game_name(new_game_name);
    cout<<endl;

    // checks if game name is valid and game name doesnt exists
    while ((games_file_process().check_valid_game_name(new_game_name) == false ||
              games_file_process().check_game_already_exists(all_games, new_game_name) ||
              games_file_process().check_game_already_exists(games_to_add, new_game_name)) &&
             new_game_name != constants::EXIT_MENU_OPTION)
    {
        new_game_name = "";
        cout << "Invalid Game Name or Game already Exists" << endl;
        cout << "Enter game name or -1 to exit: ";
        getline(cin, new_game_name);
        if (new_game_name != constants::EXIT_MENU_OPTION)
        {
            new_game_name = utils().pad_game_name(new_game_name);
        }

  }



  // getting price
  bool val_price=false;
  string text_new_game_price="";
  float  num_new_game_price=0.0;

  while(val_price==false && text_new_game_price!=constants::EXIT_MENU_OPTION && new_game_name != constants::EXIT_MENU_OPTION)
  {
    if (text_new_game_price==constants::ERROR_CODE)
    {
      cout<<"Invlaid Game Price"<<endl;
      text_new_game_price="";
    }

    cout<<"Enter Game price: ";
    cin>>text_new_game_price;
    cout<<endl;


    try {
        num_new_game_price =stof(text_new_game_price);

        if((num_new_game_price>0) && (num_new_game_price<=constants::MAX_GAME_PRICE))
        {
          val_price=true;
        }
        else
        {
          cout<<"Invlaid Game Price"<<endl;
          cout<<"Enter Game price or -1 to exit: "<<endl;

        }

    } catch (const std::invalid_argument& e) {

        text_new_game_price=constants::ERROR_CODE;

    } catch (const std::out_of_range& e) {
        text_new_game_price=constants::ERROR_CODE;
    }
  }



  if (text_new_game_price!=constants::EXIT_MENU_OPTION&&new_game_name != constants::EXIT_MENU_OPTION)
  {
    text_new_game_price=utils().pad_game_price(num_new_game_price);

    string log_entry;

    string new_game_to_sell=new_game_name+" "+current_user_name+" "+ text_new_game_price;

    log_entry= constants::SELL_GAME_CODE+" "+new_game_to_sell;
    utils().update_transction_log(log_entry,transaction_log);

    return new_game_to_sell;
  }


    return constants::FAIL_OPTION;
}

// logic for create transaction
string transactions::process_create(vector<string>& all_users, vector<string>& transaction_log){

  unordered_map <string, string> user_type_codes = {{"1","AA"},{"2","FS"},{"3","BS"},{"4","SS"}};

  string new_account;
  cout << "Create" << endl;

  cout << "Select type of account you want to create" << endl;

  menus().display_all_user_types();

  cout << "Enter account type: ";
  cin >> new_account;


  bool valid_account = false;

  int num = stoi(new_account);

  while(!valid_account){

    try {
      if(stoi(new_account) >= 1 && stoi(new_account) <= 4 ){
        break;

      }

      cout << "Please provide valid account type or -1 to go back to main menu " <<  endl;
      cout << "Enter account type: ";
      cin >> new_account;

      if (new_account == constants::EXIT_MENU_OPTION){
        return constants::EXIT_MENU_OPTION;
      }
    }

    catch (const std::invalid_argument& e) {
      cerr << "Invalid argument: " << e.what() << endl;
      cout << "Please provide valid account type: ";
      cin >> new_account;
    }

  }

  cout << "Account type selected" << endl;

  string username;
  cout << "Enter Username: ";
  cin >> username;

  username = utils().pad_username(username);


  while(user_file_process().check_user_names(all_users,username) || username.size() > constants::MAX_USER_NAME_LENGTH){

    if(user_file_process().check_user_names(all_users,username)){
      cout << "Username already exists please try again or -1 to exit" << endl;
      cout << "Enter valid Username: ";
    }

    else{
      cout << "Username length is too long please try again or -1 to exit" << endl;
      cout << "Enter valid Username: ";
    }
    cin >> username;


    if (username == constants::EXIT_MENU_OPTION){
      return constants::EXIT_MENU_OPTION;
    }

    username = utils().pad_username(username);
  }

  cout << "username selected" << endl;

  double credit_amount=0.0;
  string text_credit_amount;
  bool valid_credit_amount = false;

  cout << "Enter new credit amount: ";
  cin >> text_credit_amount;

  while(!valid_credit_amount){

    try{

      credit_amount = stod(text_credit_amount);

      if( credit_amount >= constants::MIN_CREDIT_AMOUNT && credit_amount <= constants::MAX_CREDIT_AMOUNT){
        cout << "Succesfully created new user " << endl;
        text_credit_amount = utils().pad_credit_amount(credit_amount);
        string log_entry = constants::CREATE_CODE+" "+username+ " " + user_type_codes[new_account] + " " + text_credit_amount;
        utils().update_transction_log(log_entry,transaction_log);
        return constants:: SUCESS_OPTION;
      }

      cout << "Please enter a valid credit amount or -1 to go back to main menu" << endl;
      cout << "Enter credit amount: ";
      cin >> text_credit_amount;


      if (text_credit_amount == constants::EXIT_MENU_OPTION){
        return constants::EXIT_MENU_OPTION;
      }

    }

    catch (const std::invalid_argument& e) {
      cerr << "Invalid argument: " << e.what() << endl;
      cout << "Please enter a valid credit amount or -1 to go back to main menu" << endl;
      cout << "Enter credit amount: ";
      cin>>text_credit_amount;

      if (text_credit_amount == constants::EXIT_MENU_OPTION){
        return constants::EXIT_MENU_OPTION;
      }

    }

  }


  return constants:: SUCESS_OPTION;


}

// logic for delete transactions
string transactions:: process_delete(vector<string>& all_users, string current_user,vector<string>& transaction_log){

  string user_to_remove;

  cout << "Delete" << endl;

  cout << "Enter username: ";

  cin >> user_to_remove;

  user_to_remove = utils().pad_username(user_to_remove);


  while(!user_file_process().check_user_names(all_users,user_to_remove) || user_to_remove==current_user){

    if(user_to_remove == current_user){
      cout << "user cannot remove itself please try again or -1 to exit" << endl;
      cout << "Enter username: ";
    }

    else{
      cout << "Username does not exist please try again or -1 to exit " << endl;
      cout << "Enter username: ";
    }

    cin >> user_to_remove;

    if (user_to_remove == constants::EXIT_MENU_OPTION){
      return constants::EXIT_MENU_OPTION;
    }

    user_to_remove = utils().pad_username(user_to_remove);

  }

  cout << "User sucessfully removed" << endl; 

  float user_amount = user_file_process().get_user_balance(all_users, user_to_remove);

  string log_entry = constants::DELETE_CODE+" "+user_to_remove+ " " + user_file_process().get_user_type(all_users, user_to_remove) + " " + to_string(user_amount);
  utils().update_transction_log(log_entry,transaction_log);

  return constants:: SUCESS_OPTION;;

}

// logic for show all users transaction
void transactions::show_all_users(vector<string>& all_users)
{
  cout<<"All Users"<<endl;
  cout << setw(14) << left << "User"
         << setw(5) << left << "Type" << "Balance"
         << endl;
  for(const auto& user: all_users ){

    cout<<user<< endl;

  }
}

// logic for show all game info transaction
void transactions::show_all_game_info(vector<string>& all_games)
{
  cout<<"All Games Available"<<endl;
  cout << setw(26) << left << "Game Name"
         << setw(16) << left << "Seller" << "Price"
         << endl;
  for(const auto& game: all_games ){

    cout<<game<< endl;

  }
}

// logic for refund transaction
string transactions:: refund(vector<string>& all_users,vector<string>& all_games,vector<string>& game_collec,vector<string>& transaction_log)
  {

    utils utility;
    cout<<"Refund"<<endl;
    string refund_game_name;

    cout<<"Enter Game name: ";
    cin.ignore();
    getline(cin,refund_game_name);
    refund_game_name=utility.pad_game_name(refund_game_name);
    cout<<endl;

    while (games_file_process().check_game_already_exists(all_games,refund_game_name)==false)
    {
      cout<<"Invalid game name or Game name doesn't exist"<<endl;
      cout<<"Enter Game name or -1 to quit: ";
      getline(cin,refund_game_name);

      if (refund_game_name!=constants::EXIT_MENU_OPTION)
      {

        refund_game_name=utils().pad_game_name(refund_game_name);
      }
      else
      {
        return constants::FAIL_OPTION;
      }
    }

    string seller_name;
    cout<<"Enter seller name: ";
    cin>>seller_name;
    cout<<endl;
    while(games_file_process().check_user_sells_game(all_games,refund_game_name,seller_name)==false)
    {
      cout<<"Invalid Seller name or Seller doesn't Sell Game"<<endl;
      cout<<"Enter Seller name or -1 to quit: ";
      cin>>seller_name;
      cout<<endl;

      if (seller_name==constants::EXIT_MENU_OPTION)
      {
        return constants::FAIL_OPTION;
      }

    }

    string buyer_name;
    cout<<"Enter buyer name: ";
    cin>>buyer_name;
    cout<<endl;
    while((games_file_process().check_user_owns_game(game_collec,refund_game_name,buyer_name)==false) ||
        utils().convert_to_lower(buyer_name)==utils().convert_to_lower(seller_name))
    {
      cout<<"Invalid Buyer name or Buyer doesn't Own Game"<<endl;
      cout<<"Enter Buyer name or -1 to quit: ";
      cin>>buyer_name;
      cout<<endl;


      if (buyer_name==constants::EXIT_MENU_OPTION)
      {
        return constants::FAIL_OPTION;
      }

    }

    string text_refund_price;
    float  num_refund_price=0.0;
    bool val_price=false;

    while(val_price==false && text_refund_price!=constants::EXIT_MENU_OPTION)
    {

      if (text_refund_price==constants::ERROR_CODE)
      {
        cout << "Invalid argument: "<< endl;
        cout<<"Enter Refund price or -1 to exit: "<<endl;
        text_refund_price="";

      }

      cout<<"Enter Refund Amount: ";
      cin>>text_refund_price;
      cout<<endl;

      if (text_refund_price==constants::EXIT_MENU_OPTION)
      {
        return constants::FAIL_OPTION;
      }

      try {
          num_refund_price =stof(text_refund_price);

          if(num_refund_price>0 && games_file_process().is_game_price(all_games,refund_game_name,num_refund_price))
          {
            val_price=true;
          }
          else
          {
            cout<<"Invlaid Refund Price or Refund price entered is not equal to game amount"<<endl;
            cout<<"Enter Refund price or -1 to exit: "<<endl;



          }

      } catch (const std::invalid_argument& e) {


          text_refund_price=constants::ERROR_CODE;


      } catch (const std::out_of_range& e) {
          cerr << "Out of range: " << e.what() << endl;
          text_refund_price=constants::ERROR_CODE;
      }
    }

    // proccessing the refund

    float seller_credit_balance=user_file_process().get_user_balance(all_users,seller_name);
    float buyer_credit_balance=user_file_process().get_user_balance(all_users,buyer_name);


    if ((seller_credit_balance-num_refund_price)<0)
    {

      return constants::FAIL_OPTION;

    }

    else
    {

      seller_credit_balance=seller_credit_balance-num_refund_price;
      buyer_credit_balance=buyer_credit_balance+num_refund_price;

      // DON'T UPDATE ANY BALANCES
      // ONLY WRITE TO DAILY TRANSACTION FILE

      // games_file_process().remove_game_from_user_collection(game_collec,refund_game_name,buyer_name);
      // user_file_process().update_user_balance(all_users,seller_name,seller_credit_balance);
      // user_file_process().update_user_balance(all_users,buyer_name,buyer_credit_balance);

      string log_entry;
      string log_seller_name=utils().pad_username(seller_name);
      string log_buyer_name=utils().pad_username(buyer_name);
      string log_text_refund_price=utils().pad_credit_amount(num_refund_price);

      log_entry=constants::REFUND_CODE+" "+log_buyer_name+" "+log_seller_name+" "+log_text_refund_price;
      utils().update_transction_log(log_entry,transaction_log);
    }


    return constants:: SUCESS_OPTION;


  }

// logic for buy transaction
string transactions::process_buy(string buyer_name, vector<string>& all_users, vector<string>& all_games, vector<string>& game_collec, vector<string>& transaction_log)
{
    utils utility;

    cout<<"Buy"<<endl;
    string game_name;

    cout<<"Enter Game name: ";
    cin.ignore();
    getline(cin, game_name);
    game_name = utility.pad_game_name(game_name);

    // check if game exists
    while (games_file_process().check_game_already_exists(all_games, game_name) == false)
    {
      cout << "Invalid game name or Game name doesn't exist" << endl;
      cout << "Enter Game name or -1 to quit: ";
      getline(cin, game_name);

      if (game_name != constants::EXIT_MENU_OPTION)
      {
        game_name = utils().pad_game_name(game_name);
      }
      else
      {
        return constants::FAIL_OPTION;
      }
    }

    // check if buyer already owns the requested game
    while (games_file_process().check_user_owns_game(game_collec, game_name, buyer_name) == true)
    {
      cout << buyer_name << " already owns " << game_name <<endl;
      cout << "Enter another game or enter -1 to quit: ";
      getline(cin, game_name);

      if (game_name != constants :: EXIT_MENU_OPTION)
      {
        game_name = utils().pad_game_name(game_name);
      }
      else
      {
        return constants::FAIL_OPTION;
      }
    }


    string seller_name;
    cout << "Enter seller name: ";
    cin >> seller_name;
    cout << endl;

    // check if seller sells the requested game
    while(games_file_process().check_user_sells_game(all_games, game_name, seller_name) == false)
    {
      cout << "Invalid Seller name or Seller doesn't Sell Game"<<endl;
      cout << "Enter Seller name or -1 to quit: ";
      cin >> seller_name;
      cout << endl;

      if (seller_name == constants :: EXIT_MENU_OPTION)
      {
        return constants::FAIL_OPTION;
      }
    }

    // update buyer and seller balances
    float seller_credit_balance=user_file_process().get_user_balance(all_users, seller_name);
    float buyer_credit_balance=user_file_process().get_user_balance(all_users, buyer_name);

    cout<< "BALANCE BEFORE PURCHASE" << endl;
    cout<< "SELLER " << seller_credit_balance << endl;
    cout<< "BUYER " << buyer_credit_balance << endl;
    cout << endl;

    game_name = utils().convert_to_lower(game_name);
    float game_price = games_file_process().get_game_price(all_games, game_name);

    if (game_price > buyer_credit_balance)
    {
      cout << "You do not have enough credits to buy this game. You will return to the previous menu." << endl;

      return constants :: FAIL_OPTION;
    }
    else
    {
      buyer_credit_balance = buyer_credit_balance - game_price;
      seller_credit_balance = seller_credit_balance + game_price;

      user_file_process().update_user_balance(all_users, buyer_name, buyer_credit_balance);
      user_file_process().update_user_balance(all_users, seller_name, seller_credit_balance);

      // cout << "GAMES COLLECTION BEFORE BUY TRANSACTION" << endl;
      // for (int i = 0; i < game_collec.size(); i++)
      // {
      //   cout << game_collec[i] << endl;
      // }
      // cout << endl;

      games_file_process().update_games_collection(game_collec, game_name, buyer_name);

      // cout << "GAMES COLLECTION AFTER BUY TRANSACTION" << endl;
      // for (int i = 0; i < game_collec.size(); i++)
      // {
      //   cout << game_collec[i] << endl;
      // }
      // cout << endl;

      cout<< "BALANCE AFTER PURCHASE" << endl;
      cout<< "SELLER " << seller_credit_balance << endl;
      cout<< "BUYER " << buyer_credit_balance << endl;
      // cout << endl;


      // update daily transaction file
      string log_entry;
      string log_game_name = utils().pad_game_name(game_name);
      string log_seller_name = utils().pad_username(seller_name);
      string log_buyer_name = utils().pad_username(buyer_name);
      string log_text_buy_price = utils().pad_credit_amount(game_price);

      log_entry = constants::BUY_CODE + " " + log_game_name + " " + log_seller_name + " " + log_buyer_name + " " + log_text_buy_price;
      utils().update_transction_log(log_entry, transaction_log);
    }

    return constants:: SUCESS_OPTION;
}

// process for updating a credit amount for a specific user
string transactions::process_credit(string menu_option,vector<string>& all_users, string user,vector<string>& transaction_log){

  cout << "Add Credit" << endl;

  if(menu_option == "6"){

    cout << "Enter username: ";

    cin >> user;

    user = utils().pad_username(user);

    while(!user_file_process().check_user_names(all_users,user)){

      cout << "Username does not exist please try again or -1 to exit " << endl;
      cout << "Enter username: ";

      cin >> user;


      if (user == constants::EXIT_MENU_OPTION){
        return constants::EXIT_MENU_OPTION;
      }

      user = utils().pad_username(user);

    }
  }

  cout << "username selected" << endl;

  cout << "Enter credit amount to add: ";

  float credit_amount=0.0;
  string text_credit_amount;

  cin >> text_credit_amount;

  float user_balance=user_file_process().get_user_balance(all_users, user);

  bool valid_credit_amount = false;


  while(!valid_credit_amount){

    try{

      credit_amount = stof(text_credit_amount);

      if(user_balance + credit_amount > constants::MAX_CREDIT_AMOUNT){
        cout << "credit amount added wil exceed maximum limit of $999,999 please pick appropriate credit amount to add or -1 to exit " <<  endl;
        cout << "Enter credit amount to add: ";

      }

      else if(credit_amount < constants::MIN_ADD_CREDITS || credit_amount  > constants::MAX_ADD_CREDITS) {
        cout << "Please enter a valid credit amount or -1 to go back to main menu " << endl;
        cout << "Enter credit amount to add: ";
      }

      else{
        break;
      }

      cin >> text_credit_amount;


      if (text_credit_amount == constants::EXIT_MENU_OPTION){
        return constants::EXIT_MENU_OPTION;
      }
    }

    catch (const std::invalid_argument& e) {
      cerr << "Invalid argument: " << e.what() << endl;
      cout<<"Please enter a valid credit amount or -1 to go back to main menu " << endl;
      cout << "Enter credit amount to add: ";
      cin>>text_credit_amount;
      if (text_credit_amount == constants::EXIT_MENU_OPTION){
        return constants::EXIT_MENU_OPTION;
      }
    }

  }

  user_balance+=credit_amount;

  string user_balance_padded = utils().pad_credit_amount(user_balance);

  string log_entry = constants::ADD_CREDIT_CODE+" "+user+ " " + user_file_process().get_user_type(all_users, user) + " " + user_balance_padded;
  utils().update_transction_log(log_entry,transaction_log);


  return constants:: SUCESS_OPTION;
}


void transactions::process_logout(vector<string>& transaction_log)
{
  ofstream output_stream;
  output_stream.open("log_files/daily_transactions.etf");

  for (const string& transaction : transaction_log) {
       output_stream << transaction << std::endl;
   }


  output_stream.close();


  return;
}

File: phase_6/back_end/transactions.py
from constants import Constants
import string
import utils


def process_create(transaction_line: string, current_users: list[str]):
    """
    Process a 'create' transaction.

    Parameters:
        transaction_line (string): The transaction line to process.
        current_users (list[str]): The list of current users.

    """
    start_index = Constants.MAX_ACCOUNT_TYPE_LENGTH + 1
    new_user = transaction_line[start_index:Constants.MAX_USER_NAME_LENGTH + start_index]
    new_user_code = transaction_line[19:21]
    new_user_amount = transaction_line[len(transaction_line) - Constants.MAX_USER_NAME_LENGTH + start_index + 3:len(transaction_line)]

    for user in current_users:
        if new_user in user:
            return
    current_users.append(new_user + " " + new_user_code + " " + new_user_amount)


def process_delete(transaction_line: string, current_users: list[str], available_games: list[str], games_collection: list[str]):
    """
    Process a 'delete' transaction.

    Parameters:
        transaction_line (string): The transaction line to process.
        current_users (list[str]): The list of current users.
        available_games (list[str]): The list of available games.
        games_collection (list[str]): The list of games in the collection.
    """
    start_index = Constants.MAX_ACCOUNT_TYPE_LENGTH + 1
    removed_user = transaction_line[start_index:Constants.MAX_USER_NAME_LENGTH + start_index]

    for user in current_users:
        if removed_user in user:
            utils.remove_item(current_users, removed_user)
            utils.remove_item(available_games, removed_user)
            utils.remove_item(games_collection, removed_user)
            return
    return "User not found"


def process_credit(transaction_line: string, current_users: list[str]):
    """
    Process a 'credit' transaction.

    Parameters:
        transaction_line (string): The transaction line to process.
        current_users (list[str]): The list of current users.
    """
    start_index = Constants.MAX_ACCOUNT_TYPE_LENGTH + 1
    new_user_amount = float(transaction_line[len(transaction_line) - Constants.MAX_USER_NAME_LENGTH + start_index + 3:].lstrip('0'))
    curr_user = transaction_line[start_index:Constants.MAX_USER_NAME_LENGTH + start_index]
    for i, user in enumerate(current_users):
        if curr_user in user:
            current_users[i] = utils.update_balance(user, new_user_amount)
            return
    return "User not found"


def process_buy(transaction_line: string, current_users: list[str], games_collection: list[str]):
    """
    Process a 'buy' transaction.

    Parameters:
        transaction_line (string): The transaction line to process.
        current_users (list[str]): The list of current users.
        games_collection (list[str]): The list of games in the collection.
    """
    start_index = Constants.MAX_ACCOUNT_TYPE_LENGTH + 1

    game_name = transaction_line[start_index:start_index + Constants.MAX_GAME_NAME_LENGTH]

    seller_name = transaction_line[start_index + Constants.MAX_GAME_NAME_LENGTH + 1:start_index + Constants.MAX_GAME_NAME_LENGTH + 1 + Constants.MAX_USER_NAME_LENGTH]

    buyer_name = transaction_line[start_index + Constants.MAX_GAME_NAME_LENGTH + Constants.MAX_USER_NAME_LENGTH + 2:
                                  start_index + Constants.MAX_GAME_NAME_LENGTH + Constants.MAX_USER_NAME_LENGTH + 2 + Constants.MAX_USER_NAME_LENGTH]

    game_price_current_users = transaction_line[start_index + Constants.MAX_GAME_NAME_LENGTH + 1 + Constants.MAX_USER_NAME_LENGTH + 1 + Constants.MAX_USER_NAME_LENGTH + 1:]

    game_price_available_games = game_price_current_users[3:]
    game_price_available_games = float(game_price_available_games)


    # prevents a deleted buyer from buying a game
    user_exist = False
    for i, user in enumerate(current_users):
        current_user_name = user[:Constants.MAX_USER_NAME_LENGTH]
        if  current_user_name == buyer_name:
            user_exist=True
            break

    if user_exist:
        # prevents buy from occuring if it results in a negative balance for the buyer
        for i, user in enumerate(current_users):
            current_user_name = user[:Constants.MAX_USER_NAME_LENGTH]
            if  current_user_name == buyer_name:
                balance_index: int = Constants.MAX_USER_NAME_LENGTH + 1 + Constants.MAX_ACCOUNT_TYPE_LENGTH + 1

                balance: float = float(user[balance_index:])
                balance = balance - game_price_available_games

                if balance < 0:
                    utils.display_error_message("Cannot process Buy, Buyer will have negative balance")
                    return


        game_to_add = game_name + " " + buyer_name

        game_added = False
        for collection_entry in games_collection:
            if game_to_add.lower() == collection_entry.lower():
                game_added = True

        if game_added == False:
            for i, user in enumerate(current_users):
                # balance_index: int = Constants.MAX_USER_NAME_LENGTH + 1 + Constants.MAX_ACCOUNT_TYPE_LENGTH + 1
                # user_balance: float = float(user[balance_index:])
                current_user_name = user[:Constants.MAX_USER_NAME_LENGTH]
                if current_user_name == seller_name:
                    current_users[i] = utils.update_balance(user, game_price_available_games)

                elif current_user_name == buyer_name:
                    current_users[i] = utils.update_balance(user, -game_price_available_games)

            games_collection.append(game_name + " " + buyer_name)
            utils.display_success_message("Buy")
            return
        else:
            utils.display_error_message("Buy cannot be processed, buy already processed")
            return

    print("Cannot process buy transaction, buyer not in the system")




def process_sell(transaction_line: string, available_games: list[str], games_collection: list[str],current_users: list[str]):
    """
    Process a 'sell' transaction.

    Parameters:
        transaction_line (string): The transaction line to process.
        available_games (list[str]): The list of available games.
        games_collection (list[str]): The list of games in the collection.
    """
    start_index = Constants.MAX_ACCOUNT_TYPE_LENGTH + 1
    new_game_name = transaction_line[start_index:start_index + Constants.MAX_GAME_NAME_LENGTH]
    seller_name = transaction_line[start_index + Constants.MAX_GAME_NAME_LENGTH + 1:start_index + Constants.MAX_GAME_NAME_LENGTH + 1 + Constants.MAX_USER_NAME_LENGTH]
    game_price = transaction_line[start_index + Constants.MAX_GAME_NAME_LENGTH + 1 + Constants.MAX_USER_NAME_LENGTH + 1:]

    # prevents deleted users from selling games
    user_exist=False
    for i, user in enumerate(current_users):
        current_user_name = user[:Constants.MAX_USER_NAME_LENGTH]
        if  current_user_name == seller_name:
            user_exist=True
            break

    if user_exist:
        for games in available_games:
            current_game_name = games[:Constants.MAX_GAME_NAME_LENGTH]
            if current_game_name.lower() == new_game_name.lower():
                utils.display_error_message("Game is already listed for sale")
                return

        available_games.append(new_game_name + " " + seller_name + " " + game_price)
        games_collection.append(new_game_name + " " + seller_name)
        utils.display_success_message("Game listed for sale")
        return
    utils.display_error_message("Cannot list game for sale, Seller not in the system")


def process_refund(transaction_line: string, games_collection: list[str], current_users: list[str]):
    """
    Process a 'refund' transaction.

    Parameters:
        transaction_line (string): The transaction line to process.
        games_collection (list[str]): The list of games in the collection.
        current_users (list[str]): The list of current users.
    """
    start_index = Constants.MAX_ACCOUNT_TYPE_LENGTH + 1
    game_name_length = Constants.MAX_GAME_NAME_LENGTH
    user_name_length = Constants.MAX_USER_NAME_LENGTH

    refund_game_name = transaction_line[start_index:start_index + game_name_length]
    buyer_name = transaction_line[start_index + game_name_length + 1:start_index + game_name_length + 1 + user_name_length]
    seller_name = transaction_line[start_index + game_name_length + 1 + user_name_length + 1:start_index + game_name_length + 1 + user_name_length + 1 + user_name_length]
    refund_price = transaction_line[start_index + game_name_length + 1 + user_name_length + 1 + user_name_length + 1:]

    # prevents refund from occuring if it results in a negtuive balance for the user
    for i, user in enumerate(current_users):
        current_user_name = user[:Constants.MAX_USER_NAME_LENGTH]
        if  current_user_name.lower() == seller_name.lower():
            balance_index: int = Constants.MAX_USER_NAME_LENGTH + 1 + Constants.MAX_ACCOUNT_TYPE_LENGTH + 1
            balance: float = float(user[balance_index:])
            balance =balance- float(refund_price)
            if balance<0:
                utils.display_error_message("Cannot process Refund, Seller will have negative balance")
                return


    game_to_remove = refund_game_name + " " + buyer_name
    game_removed = False
    for collection_entry in games_collection:
        if game_to_remove.lower() == collection_entry.lower():
            games_collection.remove(collection_entry)
            #print(collection_entry)
            game_removed = True

    if game_removed:
        refund_price = float(refund_price)

        for i, user in enumerate(current_users):
            current_user_name = user[:Constants.MAX_USER_NAME_LENGTH]
            if current_user_name.lower() == buyer_name.lower():
                current_users[i] = utils.update_balance(user, refund_price)

            elif current_user_name.lower() == seller_name.lower():
                current_users[i] = utils.update_balance(user, -refund_price)
        utils.display_success_message("Refund")
    else:
        utils.display_error_message("Refund cannot be processed, refund already processed or game does not exist")
